clippy_utils/src/lib.rs

```diff
@@ -586,14 +586,19 @@ pub fn def_path_res(cx: &LateContext<'_>, path: &[&str], namespace_hint: Option<
         }
     }
 
-    fn find_crate(tcx: TyCtxt<'_>, name: &str) -> Option<DefId> {
-        tcx.crates(())
+    fn find_crate(cx: &LateContext<'_>, name: &str) -> Option<DefId> {
+        cx.tcx
+            .crates(())
             .iter()
             .copied()
-            .find(|&num| tcx.crate_name(num).as_str() == name)
+            .find(|&num| cx.tcx.crate_name(num).as_str() == name)
             .map(CrateNum::as_def_id)
     }
 
+    if path.first() == Some(&"crate") || path.first().copied() == cx.sess().opts.crate_name.as_deref() {
+        return def_path_res_local(cx, &path[1..]);
+    }
```

Comment:
We may actually not want to handle `crate`, it would resolve to different paths for each package in the workspace, and I think also each bin/lib target within a single package

Could also move this down one statement to use `base` in place of `path.first()` to guard against any zero length path weirdness

Reply:
Ok, I have removed the `crate`, and used `base`.